### PR TITLE
misc(query): make spreadFunc serializable

### PR DESCRIPTION
**Pull request by veermaha**
Motivation:
  - `spreadFunc` which part of a QueryCommand needs to be serializable.

Modifications:
 - let `spreadFunc` be derived using a static provider allows it to be serializable.

Bonus:
- add a commit message template

**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?

### CODE REVIEW

File: coordinator/src/main/scala/filodb.coordinator/NodeCoordinatorActor.scala [hunk 1]
```diff
@@ -55,13 +57,21 @@ private[filodb] final class NodeCoordinatorActor(metaStore: MetaStore,
 
   val settings = new FilodbSettings(config)
   val ingesters = new HashMap[DatasetRef, ActorRef]
-  val queryActors = new HashMap[DatasetRef, ActorRef]
   var clusterActor: Option[ActorRef] = None
   val shardMaps = new HashMap[DatasetRef, ShardMapper]
   var statusActor: Option[ActorRef] = None
+  val cluster = Cluster(context.system)
```
Comment (vishramachandran) [inline coordinator/src/main/scala/filodb.coordinator/NodeCoordinatorActor.scala]: @veermaha Lets discuss about this line. We shouldnt need to instantiate the cluster here.

**Comment by veermaha** [comment]
## AFTER: When Query hits a spare node
```
[2019-04-05 11:57:40,136] DEBUG filo-standalone-akka.actor.default-dispatcher-15 f.c.TempQueryRouter [] - Member up updated list: ArrayBuffer(Actor[akka.tcp://filo-standalone@127.0.0.1:61421/user/node/coordinator#-1238428628], Actor[akka.tcp://filo-standalone@127.0.0.1:2552/user/node/coordinator#1089549943], Actor[akka://filo-standalone/user/node/coordinator#547127261])
[2019-04-05 11:57:40,144] INFO  filo-standalone-akka.actor.default-dispatcher-15 a.c.s.ClusterSingletonManager [akka.tcp://filo-standalone@127.0.0.1:61519/user/node/nodecluster] - ClusterSingletonManager state change [Start -> Younger]

[2019-04-05 11:58:12,386] INFO  filo-standalone-akka.actor.default-dispatcher-15 f.c.TempQueryRouter [] - finding query actor for dataset: prometheus from queryActors: Map()
[2019-04-05 11:58:12,386] INFO  filo-standalone-akka.actor.default-dispatcher-15 f.c.TempQueryRouter [] - returning random: Actor[akka.tcp://filo-standalone@127.0.0.1:61421/user/node/coordinator#-1238428628]
```

filo_pr304_node3(spare)
filo_pr304_node2(live)
filo_pr304_node1(live)

File: coordinator/src/main/scala/filodb.coordinator/NodeCoordinatorActor.scala [hunk 2]
```diff
@@ -233,9 +251,9 @@ private[filodb] final class NodeCoordinatorActor(metaStore: MetaStore,
 
   private def reset(origin: ActorRef): Unit = {
     ingesters.values.foreach(_ ! PoisonPill)
-    queryActors.values.foreach(_ ! PoisonPill)
+    tempQueryRouter.queryActors.values.foreach(_ ! PoisonPill)
     ingesters.clear()
-    queryActors.clear()
+    tempQueryRouter.queryActors.clear()
```
Comment (vishramachandran) [inline coordinator/src/main/scala/filodb.coordinator/NodeCoordinatorActor.scala]: We should not do this. Only the parent needs to stop the child actors.

File: coordinator/src/test/scala/filodb.coordinator/NodeCoordinatorActorSpec.scala [hunk 3]
```diff
@@ -141,7 +141,8 @@ class NodeCoordinatorActorSpec extends ActorTest(NodeCoordinatorActorSpec.getNew
       dataset1.ref
     }
 
-    it("should return UnknownDataset if attempting to query before ingestion set up") {
+    // Disabling this test as querying for unknown dataset will keep bouncing in the current setup.
+    ignore("should return UnknownDataset if attempting to query before ingestion set up") {
```
Comment (vishramachandran) [inline coordinator/src/test/scala/filodb.coordinator/NodeCoordinatorActorSpec.scala]: Can we leave a comment in TempQueryRouter to enable this test back again when workaround is removed?

File: coordinator/src/main/scala/filodb.coordinator/NodeCoordinatorActor.scala [hunk 4]
```diff
@@ -173,11 +186,10 @@ private[filodb] final class NodeCoordinatorActor(metaStore: MetaStore,
   def queryHandlers: Receive = LoggingReceive {
     case q: QueryCommand =>
       val originator = sender()
-      withQueryActor(originator, q.dataset) { _.tell(q, originator) }
+      tempQueryRouter.withQueryActor(originator, q.dataset) { _.forward(q) }
```
Comment (vishramachandran) [inline coordinator/src/main/scala/filodb.coordinator/NodeCoordinatorActor.scala]: Can we reject the message if the q.submitTime is more than 5 seconds delayed ?
This way we can stop the indefinite forward.

File: coordinator/src/main/scala/filodb.coordinator/NodeCoordinatorActor.scala (hunk 2 above)
Comment (veermaha) [reply coordinator/src/main/scala/filodb.coordinator/NodeCoordinatorActor.scala]: Fixed in 7eec2ee

File: coordinator/src/test/scala/filodb.coordinator/NodeCoordinatorActorSpec.scala (hunk 3 above)
Comment (veermaha) [reply coordinator/src/test/scala/filodb.coordinator/NodeCoordinatorActorSpec.scala]: Done in 7eec2ee

File: coordinator/src/main/scala/filodb.coordinator/NodeCoordinatorActor.scala (hunk 4 above)
Comment (veermaha) [reply coordinator/src/main/scala/filodb.coordinator/NodeCoordinatorActor.scala]: Done in 7eec2ee

**Comment by veermaha** [comment]
@vishramachandran I've made this PR address only the `spreadFunc` changes as we discussed.

File: coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala [hunk 5]
```diff
@@ -35,19 +35,40 @@ object QueryCommands {
 
   final case class SpreadChange(time: Long = 0L, spread: Int = 1)
 
+  trait SpreadProvider {
+    def spreadFunc(filter: Seq[ColumnFilter]): scala.Seq[SpreadChange]
```
Comment (velvia) [inline coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala]: Please remove the `scala.` prefix.  That is annoying IntelliJ auto insert stuff which is not necessary.

File: coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala [hunk 6]
```diff
@@ -35,19 +35,40 @@ object QueryCommands {
...
+
+  case class StaticSpreadProvider(spreadChange: SpreadChange = SpreadChange()) extends SpreadProvider {
+    def spreadFunc(filter: Seq[ColumnFilter]): scala.Seq[SpreadChange] = {
+      Seq(spreadChange)
+    }
+  }
+
+  /**
+    * Serialize with care! would be based on the provided function.
+    * @param f
```
Comment (velvia) [inline coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala]: add description here?

File: coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala [hunk 7]
```diff
@@ -35,19 +35,40 @@ object QueryCommands {
 
   final case class SpreadChange(time: Long = 0L, spread: Int = 1)
 
+  trait SpreadProvider {
+    def spreadFunc(filter: Seq[ColumnFilter]): scala.Seq[SpreadChange]
+  }
+
+  case class StaticSpreadProvider(spreadChange: SpreadChange = SpreadChange()) extends SpreadProvider {
```
Comment (velvia) [inline coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala]: final case class

Comment (veermaha) [reply coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala]: fixed in 5843940

File: coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala (hunk 6 above)
Comment (veermaha) [reply coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala]: fixed in 5843940

File: coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala (hunk 5 above)
Comment (veermaha) [reply coordinator/src/main/scala/filodb.coordinator/client/QueryCommands.scala]: fixed in 5843940

**Pull request merged**